### PR TITLE
Move scheduler plugin set and configuration defaulting to CC

### LINKED ISSUE #102211: Move scheduler plugin defaulting to ComponentConfig defaulting logic
Currently if the user sets a plugin configuration, parsing and defaulting for unset parameters for a plugin config is properly handled during [conversion](https://github.com/kubernetes/kubernetes/blob/fed3a4520d99a2d99fd4ed0f7edf1ac79e7253e7/pkg/scheduler/apis/config/v1beta1/conversion.go#L162).

However, in the case were the user doesn't specify a plugin config at all, then generating a default plugin config happens when we [instantiate the framework](https://github.com/kubernetes/kubernetes/blob/fed3a4520d99a2d99fd4ed0f7edf1ac79e7253e7/pkg/scheduler/framework/runtime/framework.go#L421)

This is not proper, and I propose we move the latter to happen during conversion. This will ensure that the plugin configuration are properly defaulted from the beginning, and any "hacks" we needed to do down stream to handle legacy configurations (like the [extenders ignoredExtendedResources](https://github.com/kubernetes/kubernetes/blob/fed3a4520d99a2d99fd4ed0f7edf1ac79e7253e7/pkg/scheduler/factory.go#L126)) is applied to a properly defaulted configuration.

/sig scheduling

### PR DESCRIPTION
**Pull request by ahg-g**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Moves scheduler plugin and plugin config defaulting to its canonical place, component config. 

#### Which issue(s) this PR fixes:
Fixes #102211

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/apis/config/v1beta1/default_plugins.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+func applyFeatureGates(config *v1beta1.Plugins) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeCapacityPriority) {
+		config.Score.Enabled = append(config.Score.Enabled, v1beta1.Plugin{Name: names.VolumeBinding, Weight: pointer.Int32Ptr(1)})
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DefaultPodTopologySpread) {
+		// When feature is enabled, the default spreading is done by
+		// PodTopologySpread plugin, which is enabled by default.
+		klog.Infof("Registering SelectorSpread plugin")
```
Comment (yuzhiquan) [inline pkg/scheduler/apis/config/v1beta1/default_plugins.go:123]:
```suggestion
		klog.InfoS("Registering SelectorSpread plugin")
```

File: pkg/scheduler/apis/config/v1beta2/default_plugins.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+func applyFeatureGates(config *v1beta2.Plugins) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeCapacityPriority) {
+		config.Score.Enabled = append(config.Score.Enabled, v1beta2.Plugin{Name: names.VolumeBinding, Weight: pointer.Int32Ptr(1)})
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DefaultPodTopologySpread) {
+		// When feature is enabled, the default spreading is done by
+		// PodTopologySpread plugin, which is enabled by default.
+		klog.Infof("Registering SelectorSpread plugin")
```
Comment (yuzhiquan) [inline pkg/scheduler/apis/config/v1beta2/default_plugins.go:122]:
```suggestion
		klog.InfoS("Registering SelectorSpread plugin")
```

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -258,15 +254,19 @@ func New(client clientset.Interface,
 
 	metrics.Register()
 
+	klog.Error("Creating scheduler")
```
Comment (yuzhiquan) [inline pkg/scheduler/scheduler.go:257]:
```suggestion
	klog.ErrorS(nil, "Creating scheduler")
```

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -258,15 +254,19 @@ func New(client clientset.Interface,
...
-		sc, err := configurator.createFromConfig()
+		// Create the config from component config
+		sc, err := configurator.create()
 		if err != nil {
 			return nil, fmt.Errorf("couldn't create scheduler: %v", err)
 		}
 		sched = sc
 
 	} else {
+		klog.Error("Creating scheduler")
```
Comment (yuzhiquan) [inline pkg/scheduler/scheduler.go:269]:
```suggestion
		klog.ErrorS(nil, "Creating scheduler")
```

File: pkg/scheduler/apis/config/v1beta2/default_plugins.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta2/default_plugins.go:2]: 2021

File: pkg/scheduler/apis/config/v1beta1/default_plugins.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta1/default_plugins.go:2]: 2021

File: pkg/scheduler/factory.go [hunk 7]
```diff
@@ -221,6 +217,12 @@ func (c *Configurator) createFromPolicy(policy schedulerapi.Policy) (*Scheduler,
 		return nil, err
 	}
 
+	// If profiles is already set, it means the user is using both CC and policy config, error out
+	// since these configs are no longer merged and they should not be used simultaneously.
+	if c.profiles != nil {
```
Comment (alculquicondor) [inline pkg/scheduler/factory.go:222]: why not do this in options.go like before?

File: pkg/scheduler/factory.go [hunk 8]
```diff
@@ -297,14 +299,17 @@ func (c *Configurator) createFromPolicy(policy schedulerapi.Policy) (*Scheduler,
...
+	c.profiles = []schedulerapi.KubeSchedulerProfile{
+		{
+			SchedulerName: v1.DefaultSchedulerName,
+			Plugins:       &plugins,
+			PluginConfig:  pluginConfig,
+		},
 	}
 
+	// Use the latest version to set the default values for plugin configs.
+	v1beta2.ConvertToInternalPluginConfigArgs(&c.profiles[0])
```
Comment (alculquicondor) [inline pkg/scheduler/factory.go:311]: But these are all internal types already. This function does nothing, I think.

I see two alternatives:
- make sure that the pluginconfigs that the legacy registry returns include the proper defaults.
- Change the legacy registry to use external types and do a conversion.

File: pkg/scheduler/framework/runtime/framework.go [hunk 9]
```diff
@@ -335,10 +329,7 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
 			continue
 		}
 
-		args, err := getPluginArgsOrDefault(options.componentConfigVersion, pluginConfig, name)
-		if err != nil {
-			return nil, fmt.Errorf("getting args for Plugin %q: %w", name, err)
-		}
+		args := pluginConfig[name]
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:332]: what about cluster autoscaler?

Comment (ahg-g) [reply pkg/scheduler/framework/runtime/framework.go:332]: Yes, I have it on my todo list, they need to create a versioned CC, run it through the defaulting logic to get an internal type and then pass that when instantiating the framework.

File: pkg/scheduler/factory.go (hunk 8 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go:311]: This function does the plugin config defaulting.

File: pkg/scheduler/factory.go (hunk 7 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go:222]: This shouldn't really happen.

> why not do this in options.go like before?

Now profiles are getting defaulted in CC defaulting logic, so the CC that we create in options.go will always have profiles set. That is why I am resetting it to nil there if you noticed.

File: pkg/scheduler/apis/config/v1beta1/conversion.go [hunk 10]
```diff
@@ -162,7 +166,33 @@ func convertToInternalPluginConfigArgs(out *config.KubeSchedulerConfiguration) e
 			if err != nil {
 				return fmt.Errorf("converting .Profiles[%d].PluginConfig[%d].Args into internal type: %w", i, j, err)
 			}
-			out.Profiles[i].PluginConfig[j].Args = internalArgs
+			prof.PluginConfig[j].Args = internalArgs
+		}
+
+		// Append default configs for plugins that didn't have one explicitly set.
+		for _, name := range prof.Plugins.Names() {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/conversion.go:173]: yeah, I was thinking the same, but left it for now just to focus on getting this to work, let me try to move to defaults.go

File: pkg/scheduler/apis/config/testing/compatibility_test.go [hunk 11]
```diff
@@ -90,1143 +90,18 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
-				TLSConfig:        &config.ExtenderTLSConfig{Insecure: true},
-				HTTPTimeout:      metav1.Duration{Duration: 1},
-				NodeCacheCapable: true,
-				ManagedResources: []config.ExtenderManagedResource{{Name: "example.com/foo", IgnoredByScheduler: true}},
-				Ignorable:        true,
-			}},
-		},
-		{
-			name: "1.16",
+			name: "all predicates and priorities",
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/testing/compatibility_test.go:104]: Yes, they don't make sense really.

File: cmd/kube-scheduler/app/options/options.go [hunk 12]
```diff
@@ -194,18 +195,18 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
...
-
 		// use the loaded config file only, with the exception of --address and --port.
 		if err := o.CombinedInsecureServing.ApplyToFromLoadedConfig(c, &c.ComponentConfig); err != nil {
 			return err
 		}
 	}
 
+	// If the user is using the legacy policy config, clear the profiles, they will be set
+	// on scheduler instantiation based on the configurations in the policy file.
+	if c.LegacyPolicySource != nil {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply cmd/kube-scheduler/app/options/options.go:206]: but there are flags in CC that users can't set in policy.

File: pkg/scheduler/apis/config/v1beta1/defaults.go [hunk 13]
```diff
@@ -55,6 +54,12 @@ func SetDefaults_KubeSchedulerConfiguration(obj *v1beta1.KubeSchedulerConfigurat
 		obj.Profiles[0].SchedulerName = pointer.StringPtr(corev1.DefaultSchedulerName)
 	}
 
+	// Add the default set of plugins and apply the configuration.
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/defaults.go:57]: I am not sure what you mean, but here we need to apply this on all profiles: add the default set of plugins into each profiles and then apply the disabled/enabled lists.

[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/defaults.go:57]: adding a default profile when non exists is done right above this logic.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta1/defaults.go:57]: What if the user didn't set any profile? Should we add one?

[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/defaults.go:57]: adding a default profile when non exists is done right above this logic.

File: cmd/kube-scheduler/app/options/options.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply cmd/kube-scheduler/app/options/options.go:206]: updated the note.

File: pkg/scheduler/apis/config/v1beta1/defaults.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta1/defaults.go:57]: ahh, sorry :)

File: cmd/kube-scheduler/app/server_test.go [hunk 14]
```diff
@@ -244,50 +212,37 @@ profiles:
 				"--config", multiProfilesConfig,
 				"--kubeconfig", configKubeconfig,
 			},
-			wantPlugins: map[string]map[string][]kubeschedulerconfig.Plugin{
-				"profile-default-plugins": defaultPlugins,
+			wantPlugins: map[string]*config.Plugins{
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply cmd/kube-scheduler/app/server_test.go]: done.

File: pkg/scheduler/apis/config/v1beta1/conversion.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/conversion.go]: ok, it worked ok.

File: pkg/scheduler/factory.go (hunk 8 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go]: So I copied the plugin defaulting logic here. It should be fine for now, in all cases policy should have its own defaulting.

File: pkg/scheduler/scheduler_test.go [hunk 15]
```diff
@@ -132,35 +132,83 @@ func TestSchedulerCreation(t *testing.T) {
 		wantProfiles []string
 	}{
 		{
-			name:         "default scheduler",
-			wantProfiles: []string{"default-scheduler"},
-		},
-		{
-			name:         "valid out-of-tree registry",
-			opts:         []Option{WithFrameworkOutOfTreeRegistry(validRegistry)},
+			name: "valid out-of-tree registry",
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler_test.go:135]: Can we add a subtest to cover that an out-of-tree with its custom pluginArgs?

File: pkg/scheduler/apis/config/testing/defaults/defaults.go [hunk 16]
```diff
@@ -0,0 +1,227 @@
+/*
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/testing/defaults/defaults.go:1]: cyclical dependency

File: pkg/scheduler/scheduler_test.go (hunk 15 above)
Comment (ahg-g) [reply pkg/scheduler/scheduler_test.go:135]: We have in scheme_test.go.

**Comment by ahg-g** [comment]
/hold 
just in case someone lgtms by mistake

I fixed the integration tests in the last commit. The PR is in its final state.

File: pkg/scheduler/factory.go [hunk 17]
```diff
@@ -307,12 +308,61 @@ func (c *Configurator) createFromPolicy(policy schedulerapi.Policy) (*Scheduler,
...
+	for j := range prof.PluginConfig {
+		existingConfigs.Insert(prof.PluginConfig[j].Name)
+		args := prof.PluginConfig[j].Args
+		if args == nil {
+			continue
+		}
+		if _, isUnknown := args.(*runtime.Unknown); isUnknown {
+			continue
+		}
+		scheme.Default(args)
```
Comment (alculquicondor) [inline pkg/scheduler/factory.go:330]: Are you sure this works? The types produced by the legacy registry are already internal. Thus, I think this and the line below do nothing.

File: pkg/scheduler/apis/config/testing/config.go [hunk 18]
```diff
@@ -25,8 +25,7 @@ import (
 )
 
 // NewV1beta1DefaultComponentConfig creates a v1beta1 default configuration.
-func NewV1beta1DefaultComponentConfig() (*config.KubeSchedulerConfiguration, error) {
-	versionedCfg := v1beta1.KubeSchedulerConfiguration{}
+func NewV1beta1DefaultComponentConfig(versionedCfg v1beta1.KubeSchedulerConfiguration) (*config.KubeSchedulerConfiguration, error) {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/testing/config.go]: perhaps a better name is:

`V1beta1ToInternalWithDefaults`

File: cmd/kube-scheduler/app/server_test.go [hunk 19]
```diff
@@ -167,7 +167,7 @@ profiles:
 			},
 		},
 		{
-			name: "plugin config with single profile",
+			name: "default configuration",
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply cmd/kube-scheduler/app/server_test.go:170]: done.

File: pkg/scheduler/apis/config/v1beta1/defaults.go [hunk 20]
```diff
@@ -39,6 +40,76 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
...
+	// Set default plugins.
+	prof.Plugins = mergePlugins(getDefaultPlugins(), prof.Plugins)
+
+	// Set default plugin configs.
+	scheme := getPluginArgConversionScheme()
+	existingConfigs := sets.NewString()
+	for j := range prof.PluginConfig {
+		existingConfigs.Insert(prof.PluginConfig[j].Name)
+		args := prof.PluginConfig[j].Args.Object
+		if args == nil {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta1/defaults.go]: done.

File: pkg/scheduler/apis/config/v1beta2/conversion.go [hunk 21]
```diff
@@ -50,62 +49,30 @@ func Convert_v1beta2_KubeSchedulerConfiguration_To_config_KubeSchedulerConfigura
...
-		if err != nil {
-			if runtime.IsNotRegisteredError(err) {
-				// This plugin is out-of-tree or doesn't require configuration.
+func convertToInternalPluginConfigArgs(out *config.KubeSchedulerConfiguration) error {
+	scheme := GetPluginArgConversionScheme()
+	for i := range out.Profiles {
+		prof := &out.Profiles[i]
+		for j := range prof.PluginConfig {
+			args := prof.PluginConfig[j].Args
+			if args == nil {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta2/conversion.go:63]: done.

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 22]
```diff
@@ -40,6 +41,74 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
...
+		gvk := v1beta2.SchemeGroupVersion.WithKind(name + "Args")
+		args, err := scheme.New(gvk)
+		if err != nil {
+			if runtime.IsNotRegisteredError(err) {
+				// This plugin is out-of-tree or doesn't require configuration.
+				continue
+			} else {
+			}
+		}
+		scheme.Default(args)
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:97]: the tests fail if I comment this out.

File: pkg/scheduler/factory.go (hunk 17 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go]: yeah, this doesn't, but the one for the newly added ones below does. but this was the existing behavior in the now removed framework.go defaulting logic (i.e., do nothing for existing plugin configs). In any case, I think we should be able to remove this completely in the next release by removing the policy flag.

Comment (ahg-g) [reply pkg/scheduler/factory.go]: removed it and added a comment.

**Review by alculquicondor — COMMENTED** [review]
good for squash

**Comment by ahg-g** [comment]
squashed

**Pull request merged**